### PR TITLE
[spmd] separate sharding propagation with compute

### PR DESCRIPTION
**Pull request by wanchaol**
This PR separates the computation and sharding propagation, so that we could enable easier/quick operator implementation. 

TODO:
add more point-wise op tests, and find a way to map between the original op impl with the aten op impl (i.e. map `torch.add` to `torch.ops.aten.add.default`, so that we could do testing easily.

### CODE REVIEW

File: spmd/tensor/ops/pointwise_ops.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+from typing import List, Optional
+from spmd.tensor.placement_types import Shard, Replicate, _Partial, PlacementSpec
+
+[
```
Comment (fduwjj) [inline spmd/tensor/ops/pointwise_ops.py:4]: QQ, how do we come up with this list and ensure its coverage?

Comment (wanchaol) [reply spmd/tensor/ops/pointwise_ops.py:4]: It's coming from https://github.com/pytorch/pytorch/blob/fd84c458f49df2ec2801a230f332ea3730d352a9/docs/source/torch.rst#L275

It seems that it's still missing other pointwise ops as it seems not a complete list. To ensure test coverage, I guess we need to loop through the list and compare them with local tensor operations

File: spmd/tensor/api.py [hunk 2]
```diff
@@ -1,30 +1,41 @@
...
+    # class attribute that handles operator placements propagation
+    # rules, keyed by aten op name, value is propagation func
+    _op_to_rules: Dict[str, Callable] = {}
 
-    # class attribute that handles ops, all handled
-    # ops should appear in this table
+    # class attribute that handles custom registered ops, all handled
+    # custom ops should appear in this table, and overriding the default
+    # operators
     # pyre-fixme[24]: Generic type `Callable` expects 2 type parameters.
```
Comment (aazzolini) [inline spmd/tensor/api.py:37]: Could you actually fix this -- it would help document the callable's semantic via its signatrue.

File: spmd/tensor/api.py [hunk 3]
```diff
@@ -1,30 +1,41 @@
...
+
+    # class attribute that handles operator placements propagation
+    # rules, keyed by aten op name, value is propagation func
+    _op_to_rules: Dict[str, Callable] = {}
 
-    # class attribute that handles ops, all handled
-    # ops should appear in this table
+    # class attribute that handles custom registered ops, all handled
+    # custom ops should appear in this table, and overriding the default
+    # operators
```
Comment (aazzolini) [inline spmd/tensor/api.py:36]: Could you clarify what "default operators" means here, is it operators that are not covered by the _op_to_rules above?

File: spmd/tensor/api.py [hunk 4]
```diff
@@ -74,12 +86,10 @@ def __torch_function__(cls, func, types, args=(), kwargs={}):
         # if we find nn.functional name in dispatch op, dispatch to it instead,
         # this allow us to override some python level behaviors that wouldn't be
         # possible in __torch_dispatch__ level.
-        if func.__name__ in DTensor._dist_tensor_dispatch_ops:
+        if func.__name__ in DTensor._custom_dispatch_ops:
```
Comment (aazzolini) [inline spmd/tensor/api.py:89]: Could it happen that there's a torch.function with the same name as a torch dispatch function? IN that case which would take priority? also is func.__name__ the same as str(func) (in the dispatch below?)

File: spmd/tensor/dispatch.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        # placement base on the redistribute cost
+        # TODO: implement auto distribute with simple
+        # cost estimation
+        if output_placements is None:
+            # do auto distributed/boxing here
+            raise NotImplementedError("auto redistribute not implemented yet!")
+
+        # run local op computation
+        local_results = op_info.op_call(
+            *op_info.args_with_local_tensor, **op_info.kwargs_with_local_tensor
```
Comment (aazzolini) [inline spmd/tensor/dispatch.py:48]: @wanchaol I remember in our discussion we had mentioned that kwargs may need to be updated by the sharding propagation, is this something that we expect to happen in-place inside of the rules? Maybe we should clarify that , or even return a new set of kwargs from the prop. rule to make it explicit?

Comment (aazzolini) [reply spmd/tensor/dispatch.py:48]: @wanchaol Another thing -- maybe a suggestion -- there's a tradeoff of generality vs. how much you can do on the sharding propagation function depending on how much info you pass into it. 

Specifically, in order for us to be able to support ahead of time propagation, would it be a good thing if we made surre to only pass "meta" tensors or "fake" tensors into the sharidn gpropagation function , so to make sure its implementaiton would never rely on .data?

File: spmd/tensor/ops/prop_rules.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    return PlacementSpec.from_dim_map(
+        input_specs[0].mesh,
+        [dim_to_sharding[dim] for dim in output_dim],
+        pending_sums,
+    )
+
+
+def mm_prop(
+    mat1_spec: PlacementSpec, mat2_spec: PlacementSpec
+) -> Optional[PlacementSpec]:
```
Comment (aazzolini) [inline spmd/tensor/ops/prop_rules.py:72]: I think it's not hard to generalize this one , maybe you can do it here:

1) group sharding by dimension of the tensor.
2) check that the sharding-mesh-dim of dimension=1 of mat1 is the same as sharding of dimension=0 of mat2 (or that they are both replicated).
3) check that the sharding for dim=0 of mat1 and dim=1 of mat2 are _not_ on the same mesh dimension as the other ones above.

**Review by aazzolini — APPROVED** [review]

File: spmd/tensor/api.py (hunk 4 above)
Comment (wanchaol) [reply spmd/tensor/api.py:89]: torch_function and torch_dispatch func.__name__ should be different for every operator, bc torch_dispatch will append a "default" to every op while torch_function won't have it. Example difference for op `mm`:

torch function `func.__name__`: `mm`
torch function `str(func)`: `<built-in method mm of type object at 0x10a5eb380>`
torch dispatch `func.__name__`: `mm.default`
torch dispatch `str(func)`: `aten.mm.default`

we are using the first and last one. I think the last one should be proper to distinguish clearly with torch_function, where it always work on `aten` ops

File: spmd/tensor/api.py (hunk 3 above)
Comment (wanchaol) [reply spmd/tensor/api.py:36]: it's actually the operator of `_op_to_rules` + fallback, in another word: `_custom_dispatch_ops` take the highest priority. i'll update the comment.

File: spmd/tensor/ops/matrix_ops.py [hunk 7]
```diff
@@ -1,109 +1,31 @@
...
-    transposed_local_mat = local_mat.t()
-    device_mesh = self.device_mesh
 
-    new_shard_dim = 1 if mat_placement.is_shard(dim=0) else 0
-    return DTensor.from_local(
-        transposed_local_mat, device_mesh, [Shard(new_shard_dim)]
-    )
+@register_prop_rule("aten.t.default")
+def transpose_rule(op_info: OpInfo) -> Optional[PlacementSpec]:
+    return einop_prop("ij->ji", op_info.args_spec)
```
Comment (fduwjj) [inline spmd/tensor/ops/matrix_ops.py:31]: N00b question, what does "eion" here stands for?

File: test/spmd/tensor/test_matrix_ops.py [hunk 8]
```diff
@@ -50,7 +50,7 @@ def test_mm(self):
         grad_res = torch.ones(12, 16)
         grad_dist_res = distribute_tensor(grad_res, device_mesh, shard_spec)
         dist_res.backward(grad_dist_res)
-        print(mat1.grad)
+        # print(mat1.grad)
```
Comment (fduwjj) [inline test/spmd/tensor/test_matrix_ops.py:53]: Nit: remove? or make it assertTrue(mat1.grad is not None)?

File: test/spmd/tensor/test_tp_sharding_ops.py [hunk 9]
```diff
@@ -127,7 +127,6 @@ def test_replicated_permute(self):
         new_dt = dist_tensor.permute(1, 0, 2)
         self.assertTrue(new_dt.placements[0].is_replicate())
         self.assertEqual(new_dt.to_local(), tensor.permute(1, 0, 2))
-        self.assertEqual(new_dt.stride(), tensor.permute(1, 0, 2).stride())
```
Comment (fduwjj) [inline test/spmd/tensor/test_tp_sharding_ops.py:130]: Why do we remove this?

File: spmd/tensor/placement_types.py [hunk 10]
```diff
@@ -38,4 +39,46 @@ class Replicate(Placement):
...
+        # and int >=0 represent shard on that device mesh dim
+        r = [-1] * self.ndim
+        for i, placement in enumerate(self.placements):
+            if placement.is_shard():
+                shard_dim = cast(Shard, placement).dim
+                r[shard_dim] = i
+        return r
+
+    @classmethod
+    def from_dim_map(
```
Comment (fduwjj) [inline spmd/tensor/placement_types.py:66]: It would be nice that we can add more documentations on how to use this API and `dim_map`. Or if you want to merge this one first, we can add a TODO here for a detailed document which covers different scenarios on how users can use this API to define specs for examples like [Replicate(), Shard(0), Shard(1)], or [Shard(0), Shard(0)].

File: spmd/tensor/ops/prop_rules.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        ij,j->ij - broadcasted addition
+    """
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    max_dim = max(input.ndim for input in input_specs)
+    dimchars = []
+    for input in input_specs:
+        start_dim = max_dim - input.ndim
+        p = alphabet[start_dim:max_dim]
+        dimchars.append(p)
+    out_dimchars = alphabet[:max_dim]
```
Comment (fduwjj) [inline spmd/tensor/ops/prop_rules.py:118]: Why this one does not come with start_dim?

File: spmd/tensor/dispatch.py (hunk 5 above)
Comment (wanchaol) [reply spmd/tensor/dispatch.py:48]: chatted offline: We are actually only updating the “non-tensor” args or “non-tensor” kwargs, which means we don’t even need the “tensor” args/kwargs metadata (i.e. `tensor.view(*shape)`, no need for tensor metadata like size, we only need to update the *shape tuple to update the local tensor shape). The only reason we keep “tensor” args in `args_with_local_tensor`, is that after the modification on “non-tensor” args, we could directly pass `args_with_local_tensor` to the local op computation `op_info.op_call(*op_info.args_with_local_tensor, **op_info.kwargs_with_local_tensor)`. We will need to figure out a way to disallow modification of "tensors" data on `args_with_local_tensor` with each sharding propagation rules.

Comment (aazzolini) [reply spmd/tensor/dispatch.py:48]: Makes sense. Just note that we should if possible disallow _read_ access to .data as well, otherwise we won't be able to do ahead-of-time sharding propagation.

File: spmd/tensor/placement_types.py (hunk 10 above)
Comment (wanchaol) [reply spmd/tensor/placement_types.py:85]: added :)

File: spmd/tensor/ops/prop_rules.py (hunk 11 above)
Comment (wanchaol) [reply spmd/tensor/ops/prop_rules.py:119]: this is the logic to deal with implicit broadcasting, where the out_dimchars would always be the one that have the max_dim

File: test/spmd/tensor/test_tp_sharding_ops.py (hunk 9 above)
Comment (wanchaol) [reply test/spmd/tensor/test_tp_sharding_ops.py]: sorry that was a accidental delete, added back

File: spmd/tensor/ops/prop_rules.py (hunk 6 above)
Comment (wanchaol) [reply spmd/tensor/ops/prop_rules.py:72]: yeah this make sense, this PR is getting too long, let me create a follow up issue and we can update incrementally :)

Comment (wanchaol) [reply spmd/tensor/ops/prop_rules.py:72]: https://github.com/pytorch/PiPPy/issues/335

File: spmd/tensor/dispatch.py (hunk 5 above)
Comment (wanchaol) [reply spmd/tensor/dispatch.py]: I just updated the PR with OpSchema, where I explicitly document "what can be used" and "what can be changed" as propagation rule author. "tensor" argument should be explicitly disallowed, but we didn't have a way to check if it's getting modified or not. We can take this as a follow up and make sure there's no way to modify/read tensor data in propagation rules https://github.com/pytorch/PiPPy/issues/336

regarding on disallow read access of .data, could you elaborate a bit more on it, is it because ahead of time sharding propagation likely won't have any .data materialized?
